### PR TITLE
introducing custom iPXE firmware buiding support

### PR DESCRIPTION
**Pull request by Rozzii**
This PR:
  - Implements a function to initiate iPXE firmware bilding process
     based on the 'buildipxe' script of the 'ironic-image'
  - Implements logic to generate TLS certificates for the custom
     iPXE firmware
   - Adds a set of iPXE specific environment variables to help
     parameterizing the custom iPXE build process

This pr can be merged only after https://github.com/metal3-io/metal3-dev-env/pull/1272 has been merged.

### CODE REVIEW

**Comment by Rozzii** [comment]
/hold
until #1272 is merged

File: 03_launch_mgmt_cluster.sh
Comment (Rozzii) [reply 03_launch_mgmt_cluster.sh]: done

File: 02_configure_host.sh
Comment (elfosardo) [inline 02_configure_host.sh:286]: removing this is quite destructive for upstream ironic CI, especially removing IRONIC_FROM_SOURCE logic
https://opendev.org/openstack/ironic/src/branch/master/playbooks/metal3-ci/run.yaml#L12

**Review by elfosardo — CHANGES_REQUESTED** [review]
thanks Adam, I spot a potentially destructive thing that we may need to rethink

File: 02_configure_host.sh
Comment (Rozzii) [reply 02_configure_host.sh:286]: I have not removed the IRONIC_FROM_SOURCE logic, it is here:https://github.com/metal3-io/metal3-dev-env/blob/main/lib/common.sh#L190 and https://github.com/metal3-io/metal3-dev-env/pull/1273/files#diff-62500b75da29026f109be7fdfb07b414f10e88c5967ce0e9482c8d79e14e6dddL320

I have just restructured it, I have used IRONIC_FROM_SOURCE heavily for Ipxe and basic auth work.

Comment (elfosardo) [reply 02_configure_host.sh:286]: thanks for the clarification Adam, haven't checked the rest as I saw this change and got a bit worried! :)
it should be fine then, I will run a test in ironic upstream CI to double-check
great work btw

**Comment by Rozzii** [comment]
This breaks the job https://github.com/metal3-io/project-infra/blob/7bec129e275c8055fb91d8686b90476bede7ec10/jenkins/scripts/integration_test_env.sh#L62, I will fix this, I believe this has never worked as intended.

File: 02_configure_host.sh
Comment (tuminoid) [inline 02_configure_host.sh:18]:
```suggestion
  sudo rm -f  "${IRONIC_DATA_DIR}/html/images/ironic-python-agent*"
```

File: 02_configure_host.sh
Comment (tuminoid) [inline 02_configure_host.sh:21]:
```suggestion
    --directory "${IRONIC_DATA_DIR}/html/images"
```

File: 02_configure_host.sh
Comment (tuminoid) [inline 02_configure_host.sh:169]: This or next indent is wrong as they're on same level.

File: 02_configure_host.sh
Comment (tuminoid) [inline 02_configure_host.sh:16]: Just noticed that #1272 has the exact same change for this script?

File: 03_launch_mgmt_cluster.sh
Comment (tuminoid) [inline 03_launch_mgmt_cluster.sh:467]:
```suggestion
build_ipxe_firmware() {
```

File: 03_launch_mgmt_cluster.sh
Comment (tuminoid) [inline 03_launch_mgmt_cluster.sh:502]:
```suggestion
            -v "${IRONIC_DATA_DIR}":/shared --entrypoint /bin/buildipxe \
```

**Review by tuminoid — COMMENTED** [review]
Some nits (that are duplicate from #1272 and some novel ones on top.

File: 02_configure_host.sh
Comment (Rozzii) [reply 02_configure_host.sh:277]: Answered in https://github.com/metal3-io/metal3-dev-env/pull/1272

**Review by tuminoid — COMMENTED** [review]
LGTM. I want someone else also looking through it, and lgtm it.

File: 02_configure_host.sh
Comment (lentzi90) [inline 02_configure_host.sh:367]: Nice!

File: 03_launch_mgmt_cluster.sh
Comment (lentzi90) [inline 03_launch_mgmt_cluster.sh:474]: I would suggest reversing this so we check if BUILD_IPXE!=true and return early in that case.

File: 03_launch_mgmt_cluster.sh
Comment (lentzi90) [inline 03_launch_mgmt_cluster.sh:503]: Should we not use `IRONIC_IMAGE` here or something like that?

**Comment by Rozzii** [comment]
I have modified it quite a lot on my local branch since it was decided to separate the ironic-image and the ipxe-builder, I am putting a WIP label on it until I can push the updates.

File: 03_launch_mgmt_cluster.sh
Comment (Rozzii) [reply 03_launch_mgmt_cluster.sh:474]: Done

**Review by lentzi90 — COMMENTED** [review]
/approve

**Comment by Rozzii** [comment]
I have refreshed this PR and decoupled the image collision fix from this.
/test-centos-integration-release-1-5
/test-ubuntu-e2e-integration-main

File: 03_launch_mgmt_cluster.sh
Comment (tuminoid) [reply 03_launch_mgmt_cluster.sh:465]: This still have the function part.

File: 02_configure_host.sh
Comment (tuminoid) [inline 02_configure_host.sh:337]: Why two empty lines?

File: 03_launch_mgmt_cluster.sh
Comment (tuminoid) [inline 03_launch_mgmt_cluster.sh:503]:
```suggestion
        -v "${IRONIC_DATA_DIR}":/shared "${IPXE_BUILDER_IMAGE}"
```

**Review by tuminoid — COMMENTED** [review]
3 minor nits.

**Comment by tuminoid** [comment]
You fixed another `function` but not the one you added. Other than that LGTM

**Review by tuminoid — COMMENTED** [review]
/lgtm

**Pull request merged**
